app/controllers/vouchers_controller.rb

```diff
@@ -0,0 +1,85 @@
+class VouchersController < ApplicationController
+  before_action :set_voucher, only: [:invalidate]
```

Comment:
Instead of this, I've tried to call `set_voucher` directly at the beginning of the `invalidate` method. However, the last controller test is failing in this case as the voucher gets invalidated even by the general user. Do you know why this is the case?

Reply:
The reason is pretty much the same as in the above comment: If you place the call to `set_voucher` in the invalidate method and remove the before_action, then  `@voucher` is not defined in the moment when cancancan does its job,
namely after the before actions but before the invalidate action in this case.

Reply:
See my other comment and commit cb5863dda.